executor/aggfuncs/func_group_concat.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package aggfuncs
+
+import (
+	"bytes"
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+	"strings"
+)
+
+type baseGroupConcat4String struct {
+	baseAggFunc
+}
+
+func (e *baseGroupConcat4String) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4ConcatString)(pr)
+	if p.buffer == nil {
+		chk.AppendNull(e.ordinal)
+		return nil
+	}
+	chk.AppendString(e.ordinal, p.buffer.String())
+	return nil
+}
+
+type basePartialResult4GroupConcat struct {
+	sep       string
+	sepInited bool
+	buffer    *bytes.Buffer
+}
+
+type partialResult4ConcatString struct {
+	basePartialResult4GroupConcat
+}
+
+type groupConcat4String struct {
+	baseGroupConcat4String
+}
+
+func (e *groupConcat4String) AllocPartialResult() PartialResult {
+	return PartialResult(new(partialResult4ConcatString))
+}
+
+func (e *groupConcat4String) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4ConcatString)(pr)
+	p.sep, p.sepInited, p.buffer = "", false, nil
```

Comment:
`p.sep` should be a field of `groupConcat4String`, not the partial result. The `separator` can only be a constant and should be consider as a property of the `group_concat` function.